lib/Analysis/NoiseAnalysis/BGV/NoiseCanEmbModel.h

```diff
@@ -0,0 +1,74 @@
+#ifndef INCLUDE_ANALYSIS_NOISEANALYSIS_BGV_NOISECANEMBMODEL_H_
+#define INCLUDE_ANALYSIS_NOISEANALYSIS_BGV_NOISECANEMBMODEL_H_
+
+#include <cassert>
+#include <string>
+
+#include "lib/Analysis/NoiseAnalysis/Noise.h"
+#include "lib/Parameters/BGV/Params.h"
+
+namespace mlir {
+namespace heir {
+namespace bgv {
+
+// coefficient embedding noise model
+// both average-case (from HPS19/KPZ21) and worst-case
+// use template here just for the sake of code reuse
+// W for worst-case, P for public key
+template <bool P>
+class NoiseCanEmbModel {
+ public:
+  // for KPZ21, NoiseState stores the bound ||e|| for error e.
+  using StateType = NoiseState;
+  using SchemeParamType = bgv::SchemeParam;
+  using LocalParamType = bgv::LocalParam;
+
+ private:
+  static double getVarianceErr(const LocalParamType &param);
+  static double getVarianceKey(const LocalParamType &param);
+  static double getBScale(const LocalParamType &param);
+  static double getBKs(const LocalParamType &param);
+  static double getAssuranceFactor(const LocalParamType &param);
+  static double getPhi(const LocalParamType &param);
+  static double getRingExpansionFactor(const LocalParamType &param);
+
+  static StateType evalEncryptPk(const LocalParamType &param);
+  static StateType evalEncryptSk(const LocalParamType &param);
+  static StateType evalRelinearizeHYBRID(const LocalParamType &inputParam,
+                                         const StateType &input);
+
+ public:
+  static StateType evalEncrypt(const LocalParamType &param);
+  static StateType evalConstant(const LocalParamType &param);
+  static StateType evalAdd(const StateType &lhs, const StateType &rhs);
+  static StateType evalMul(const LocalParamType &resultParam,
+                           const StateType &lhs, const StateType &rhs);
+  static StateType evalRelinearize(const LocalParamType &inputParam,
+                                   const StateType &input);
+  static StateType evalModReduce(const LocalParamType &inputParam,
+                                 const StateType &input);
+
+  // logTotal: log(Q / (t * 2))
+  // logBound: bound on ||e|| predicted by the model
+  // logBudget: logTotal - logBound
+  // as ||e|| < Q / (t * 2) for correct decryption
```

Comment:
This seems to be copy-pasted from B/FV code. See BGV code for reference. At least logTotal is `log(Ql / 2)`.

Reply:
Corrected the comment with the reference from the BGV code